### PR TITLE
Introduce initial support for multiple distributions.

### PR DESCRIPTION
**Pull request by fnordahl**
Please review commit by commit.

Signed-off-by: Frode Nordahl <frode.nordahl@canonical.com>

### CODE REVIEW

**Comment by igsilya** [comment]
Hi @fnordahl .  Thanks for the PR!
I didn't go too deep into review, but one thing that is needed here, I think, is CI.  I spent some time with cirrus and the following change seems to work (ubuntu images have less available disk space for some reason):
```
diff --git a/.cirrus.yml b/.cirrus.yml
index f24475b..69602d7 100644
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -1,14 +1,19 @@
 low_scale_task:
 
   compute_engine_instance:
-    image_project: fedora-cloud
-    image: family/fedora-cloud-35
+    matrix:
+      - image_project: fedora-cloud
+        image: family/fedora-cloud-35
+      - image_project: ubuntu-os-cloud
+        image: family/ubuntu-2204-lts
     platform: linux
     memory: 8G
+    disk: 20
 
   env:
-    DEPENDENCIES: git ansible ansible-collection-ansible-posix
-                  podman podman-docker
+    DEPENDENCIES: git ansible
+    FEDORA_DEP: podman podman-docker ansible-collection-ansible-posix
+    UBUNTU_DEP: docker.io
     PHYS_DEPLOYMENT: ${CIRRUS_WORKING_DIR}/physical-deployments/ci.yml
 
   runtime_cache:
@@ -23,7 +28,12 @@ low_scale_task:
     - ssh root@$(hostname) -v echo Hello
 
   install_dependencies_script:
-    - dnf install -y ${DEPENDENCIES}
+    - 'if [ $(which dnf) ]; then
+        dnf install -y ${DEPENDENCIES} ${FEDORA_DEP};
+       fi'
+    - 'if [ $(which apt) ]; then
+        apt update && apt install -y ${DEPENDENCIES} ${UBUNTU_DEP};
+       fi'
 
   unpack_caches_script:
     - mkdir -p runtime runtime-cache
```

Would be great if you can include something like that, as missing dependencies are hard to track in ovn-heater.

On the other note, your current implementation seem to require docker, while for rpm-based distros we allow using podman as well.   Do you think it makes sense to allow podman for debian-based systems as well?
We also had a lot of issues with docker in the past (mainly a requirement to be logged in and some unfortunate deamon hangs), so the question of supporting only podman and run s/docker/podman/ on a codebase was brought up several times in the past.
Do you think it makes sense to support only podman?  Most of major distributions seem to package it (I didn't check debian).

Also, if I install podman in the .cirrus.yml for ubuntu I'm getting errors later from ./do.sh install, likely because podman is getting replaced by docker.io in the process and something goes wrong:
```
TASK [Pull latest containers] **************************************************
fatal: [cirrus-task-6259513473368064]: FAILED! => {"changed": true, "cmd": "set -e\ndocker pull cirrus-task-6259513473368064:5000/ovn/ovn-multi-node\ndocker tag cirrus-task-6259513473368064:5000/ovn/ovn-multi-node ovn/ovn-multi-node\n", "delta": "0:00:00.050686", "end": "2023-03-10 11:56:42.841776", "msg": "non-zero return code", "rc": 1, "start": "2023-03-10 11:56:42.791090", "stderr": "Error response from daemon: Get \"https://cirrus-task-6259513473368064:5000/v2/\": http: server gave HTTP response to HTTPS client", "stderr_lines": ["Error response from daemon: Get \"https://cirrus-task-6259513473368064:5000/v2/\": http: server gave HTTP response to HTTPS client"], "stdout": "Using default tag: latest", "stdout_lines": ["Using default tag: latest"]}
```

**Comment by fnordahl** [comment]
Thank you for the review @igsilya and thanks a lot for contributing the Cirrus CI recipe, I included it verbatim in the PR with appropriate attribution etc. Making use of the `podman` and `podman-docker` packages appear to work fine, so happy to go with that, change incorporated in the updated version of the PR.

Cheers!

**Comment by dceara** [comment]
> Thank you for the review @igsilya and thanks a lot for contributing the Cirrus CI recipe, I included it verbatim in the PR with appropriate attribution etc.

@fnordahl according to the [contributing guidelines](https://github.com/dceara/ovn-heater/blob/main/CONTRIBUTING.md#tags) (which I just pushed last week :D) all co-authors must sign off (like in ovs/ovn).  @igsilya are you ok with me or @fnordahl adding your `Signed-off-by` to the last commit?

**Comment by igsilya** [comment]
> > Thank you for the review @igsilya and thanks a lot for contributing the Cirrus CI recipe, I included it verbatim in the PR with appropriate attribution etc.
> 
> @fnordahl according to the [contributing guidelines](https://github.com/dceara/ovn-heater/blob/main/CONTRIBUTING.md#tags) (which I just pushed last week :D) all co-authors must sign off (like in ovs/ovn). @igsilya are you ok with me or @fnordahl adding your `Signed-off-by` to the last commit?

Sure.  Though, I'd move the `podman` and `podman-docker` to `DEPENDENCIES` and removed the `UBUNTU_DEP` for a time being since we don't really need it.

File: do.sh
Comment (dceara) [inline do.sh:37]: Unfortunately distros don't make this easy.

This works on Fedora and Ubuntu but on my RHEL8 `/etc/os-release` I have:
```
ID="rhel"
ID_LIKE="fedora"
```
Which makes the above compute:
```
$ echo $DISTRO_IDS
"rhel" "fedora"
```
I changed it to
```
DISTRO_IDS=$(awk -F= '/^ID=/{print$2}' /etc/os-release | tr -d '"')
```
and this seemed to do the right thing for both Fedora and RHEL.

File: do.sh
Comment (dceara) [inline do.sh:67]: Should we say here that we currently only detect rpm-based and deb-based distros?

File: do.sh
Comment (dceara) [inline do.sh:99]: I know it was like this before your change but we (RedHat) don't use docker-ce anymore downstream.  We use podman downstream (and in CirrusCI).  I wonder if we should just remove docker-ce all together and simplify things..

File: do.sh
Comment (dceara) [inline do.sh:113]: Instead of duplicating the `die_distro` check does it make sense to just do it once, in the beginning of the `install()` function?

File: do.sh
Comment (dceara) [inline do.sh:126]: Do we actually need this at all?  We don't pull anything on the orchestrator node so can this whole block go away?  Of course, not a problem of this commit.

File: ovn-fake-multinode-utils/get-config-value.py
Comment (dceara) [inline ovn-fake-multinode-utils/get-config-value.py:1]: `ovn-fake-multinode-utils/translate_yaml.py` was recently added.  We should change python to python3 there too I guess.

**Review by dceara — CHANGES_REQUESTED** [review]
Thanks for the PR, looks good!  I just have a few (relatively minor) comments.

File: do.sh
Comment (fnordahl) [reply do.sh:37]: :+1:

File: do.sh
Comment (fnordahl) [reply do.sh:113]: In order to do this check once, we would need to change the structure of the program, i.e. we would need to install local deps first and then run the ansible playbook for other nodes afterwards, but I guess that's fine too

File: do.sh
Comment (fnordahl) [reply do.sh:99]: Sure, I'll add a separate commit for that

File: do.sh
Comment (fnordahl) [reply do.sh:67]: :+1:

File: do.sh
Comment (fnordahl) [reply do.sh]: Fixed in https://github.com/dceara/ovn-heater/pull/157/commits/d92932f6b4eae36dbd3a8f4dff61c0c709086b4b

File: do.sh
Comment (fnordahl) [reply do.sh:100]: Done

File: ovn-fake-multinode-utils/get-config-value.py
Comment (fnordahl) [reply ovn-fake-multinode-utils/get-config-value.py:1]: Done

File: do.sh
Comment (fnordahl) [reply do.sh]: Fixed in https://github.com/dceara/ovn-heater/pull/157/commits/c7af0c657747450952e8eb495cad8116fbc80f52

File: do.sh
Comment (igsilya) [inline do.sh]: I guess, we should also remove an attempt to install docker-ce from the `ovn-fake-multinode-utils/playbooks/install-dependencies.yml`.

Comment (fnordahl) [reply do.sh]: :+1:

File: do.sh
Comment (dceara) [inline do.sh:37]: This works now but still sets `DISTRO_IDS` to `rhel fedora` on my RHEL machine.

Nit: I think I'd still change this to:
 ```
DISTRO_IDS=$(awk -F= '/^ID=/{print$2}' /etc/os-release | tr -d '"')
```

But I'll leave it up to you, it's OK as is too.

File: do.sh
Comment (dceara) [reply do.sh:100]: Hmm, I'm sorry, I think my initial comment wasn't clear.  I actually meant this:
```
diff --git a/do.sh b/do.sh
index 6f00c08..b237600 100755
--- a/do.sh
+++ b/do.sh
@@ -303,6 +303,7 @@ function pull_ovn_tester() {
 }
 
 function install() {
+    ! is_rpm_based && ! is_deb_based && die_distro
     pushd ${rundir}
     install_deps
     install_venv
```

On a second thought, isn't it cleaner if we just add per distro `install_deps` functions.  E.g.:
```
function install() {
    pushd ${rundir}
    if is_rpm_based
    then
        install_deps_rpm
    elif is_deb_based
            install_deps_deb
    else
        die_distro
    fi
    install_venv
    configure_docker
    install_ovn_fake_multinode
    init_ovn_fake_multinode
    pull_ovn_fake_multinode
    install_ovn_tester
    pull_ovn_tester
    popd
}
```

We'll have to duplicate a bit of logic in the per-distro `install_deps_*` functions but, if we ensure that we test all supported distro types in CI, that might be acceptable.

What do you think?

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (dceara) [inline ovn-fake-multinode-utils/playbooks/deploy-minimal.yml:7]: Nit: what if we use the `ansible_os_family` (https://docs.ansible.com/ansible/2.6/user_guide/playbooks_conditionals.html#ansible-os-family) conditional here and different tasks per OS family?

Comment (dceara) [reply ovn-fake-multinode-utils/playbooks/deploy-minimal.yml:7]: Using `ansible_os_family` would also allow us to use a a single `install-dependencies.yml` playbook.

File: ovn-fake-multinode-utils/playbooks/install-dependencies-deb.yml
Comment (dceara) [inline ovn-fake-multinode-utils/playbooks/install-dependencies-deb.yml:1]: As mentioned above, I think we can merge this back into a single `install-dependencies.yml` where we add two conditional tasks (depending on OS family).  Wdyt?

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (dceara) [inline ovn-fake-multinode-utils/playbooks/deploy-minimal.yml:36]: Same nit as above about potentially using `ansible_os_family`.

File: do.sh
Comment (fnordahl) [reply do.sh:37]: Using the `^ID` to get both the ID and ID_LIKE fields was intentional, that allows the check functions to iterate over all the ID's to help them decide what kind of distro we're dealing with. I'll add a comment to make that more clear.

File: do.sh
Comment (fnordahl) [reply do.sh:100]: That does indeed look more appealing, makes it possible to glance at the `install` function and immediately understand what it does without all the implementation details. I'll redo this part, thx!

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/deploy-minimal.yml:7]: I was wondering if there was a simple way of doing this with Ansible, and it looks like there is, I'll have a go at it. Thx for pointing it out!

File: do.sh
Comment (fnordahl) [reply do.sh]: It turns out this *IS* required when using podman, otherwise we'll get the following while building the OVS/OVN containers which is done on the orchestrator:
```
Error: trying to reuse blob sha256:178c5a3203aea37107a1afcc9a965d6f9058cb12e31f6ecc900a48c310b27389 at destination: pinging container registry localhost:5000: Get "https://localhost:5000/v2/": http: server gave HTTP response to HTTPS client
```

I guess we can prune it to only hold the `[registries.insecure]` and `[registries.block]` sections?

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/deploy-minimal.yml]: Alternatively, maybe systemd module can correctly ignore re-starts of socket-activated services?
I.e. maybe we will not need to try to restart it twice and || true, if we split the task in two, one that updates teh config and one that restarts the daemon with systemd module.

I didn't try that though.

File: do.sh
Comment (dceara) [reply do.sh:41]: Ok, but I'm not sure I understand where the iteration is.  Did you mean something like:
```
function is_rpm_based() {
    for i in $DISTRO_IDS; do
        case $i in
        centos* | rhel* | fedora*)
            true
            return
        ;;
        esac
    done
    false
}
```

Comment (fnordahl) [reply do.sh:41]: So, I intuitively assumed the `in` operator would do both, and AFAICT it does?

```bash
#!/bin/bash

function is_abc() {
    echo "evaluatiing: '$DISTRO_IDS'"
    case $DISTRO_IDS in
        a* | b* | c*)
            echo a, b or c
        ;;
        *)
            echo neither
        ;;
    esac	
}

DISTRO_IDS="c"
is_abc
DISTRO_IDS="c b"
is_abc
DISTRO_IDS="c b
a"
is_abc
DISTRO_IDS="a b
c"
is_abc
DISTRO_IDS="d e
f"
is_abc
```

Output:
```
evaluatiing: 'c'
a, b or c
evaluatiing: 'c b'
a, b or c
evaluatiing: 'c b
a'
a, b or c
evaluatiing: 'a b
c'
a, b or c
evaluatiing: 'd e
f'
neither
```

Comment (fnordahl) [reply do.sh:41]: Done

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/deploy-minimal.yml]: Yup, I ended up doing that, and apparently the `state: reloaded` gracefully works fine with a socket activated service

File: ovn-fake-multinode-utils/playbooks/deploy-minimal.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/deploy-minimal.yml]: Done

File: ovn-fake-multinode-utils/playbooks/install-dependencies-deb.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies-deb.yml]: Done

File: do.sh
Comment (dceara) [reply do.sh:41]: Thanks!

**Review by dceara — APPROVED** [review]
Looks good to me, thanks!

**Comment by dceara** [comment]
@igsilya Do you want to have another look at these changes before I merge the PR?

**Comment by igsilya** [comment]
> @igsilya Do you want to have another look at these changes before I merge the PR?

Sure.  I can take a look.  I won't test though. :)

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (igsilya) [inline ovn-fake-multinode-utils/playbooks/install-dependencies.yml:7]: Can we use `ansible.builtin.package` module for this as well?
`ansible.builtin.dnf/yum` should also be fine, since we're in the OS-specific part here.

**Comment by igsilya** [comment]
I added 2 small comments.  Otherwise LGTM.

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:7]: The reason for not doing that was the `--skip-broken` part in the original code.

If the reason for doing that is now gone, or we can accomplish something similar with Ansible, I'm fine with changing it. Anyone know why it was there in the first place?

Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:7]: I'm not sure why it is there.  But `ansible.builtin.dnf` module has it: https://docs.ansible.com/ansible/latest/collections/ansible/builtin/dnf_module.html#parameter-skip_broken

Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:7]: Thx for the pointer!

After some light inspection, it does appear Ansible would potentially pass on task variables to whatever backend package module is used, ref: https://github.com/ansible/ansible/blob/bad8843124a50493141a3e3d7920353239021389/lib/ansible/plugins/action/package.py

I do not know for a fact that it works that way though. I'll try to add the `ansible.builtin.package` module and we'll see if that is inherited to the `dnf` module on RHEL systems?

I don't have one at hand so would appreciate if one of you would be able to check whether it works there?

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:17]: Done

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Done

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:17]: I think, you missed the `name:` part.

Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml:17]: Ugh.  Wrong thread.

File: ovn-fake-multinode-utils/playbooks/install-dependencies.yml
Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: I think, you missed the `name:` part.

Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Nah, that didn't work, trying to duplicate the task with some some nitty gritty conditionals around RH distro specifics then.

Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]:
```
Syntax Error while loading YAML.
  did not find expected '-' indicator

The error appears to be in '/tmp/cirrus-ci-build/ovn-fake-multinode-utils/playbooks/install-dependencies.yml': line 13, column 9, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

        - containers-common
        state: present
        ^ here
```

I it was the same error before.  I guess it should be:
```
 hosts: ovn_hosts
  tasks:
    - name: Install required packages
      ansible.builtin.package:
        name:
          - git
          - gcc
          - openvswitch
          - python3-pyyaml
          - python3-virtualenv
          - python3-devel
          - containers-common
        skip_broken: true
        state: present
      when: ansible_os_family == "RedHat"
```

Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: On the [original change attempt](https://cirrus-ci.com/task/6206298660798464) it specifically choked on the `skip_broken` key

Comment (igsilya) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Yes, because you didn't have a `name:` qualifier.  So, it tried to interpret `skip_broken` as part of the list.  If we have a `name:` and package list is indented, it shouldn't fail this way, AFAICT.

Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Ha, indeed, good spot, thx!

Comment (dceara) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Unfortunately 'skip-broken' is broken (pun intended) on some ansible versions:
https://github.com/ansible/ansible/issues/73072

Officially this should've been starting with ansible 2.14.3 (https://github.com/ansible/ansible/commit/6bcb494f8306615f2b8741dad23529fdcd94626c) but I didn't manage to make it work on our RHEL8.7.  I'm using 

```
# ansible --version
ansible [core 2.14.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.9.13 (main, Nov  9 2022, 13:16:24) [GCC 8.5.0 20210514 (Red Hat 8.5.0-15)] (/usr/bin/python3.9)
  jinja version = 3.1.2
  libyaml = True
```

And still getting errors for missing packages:
```
TASK [Install required packages] *************************************************************************************************************************************************************************************************************
fatal: [advnetlab29.anl.eng.rdu2.redhat.com]: FAILED! => {"changed": false, "failures": ["No package openvswitch available."], "msg": "Failed to install some of the specified packages", "rc": 1, "results": []}                            
fatal: [advnetlab11.anl.eng.rdu2.redhat.com]: FAILED! => {"changed": false, "failures": ["No package openvswitch available."], "msg": "Failed to install some of the specified packages", "rc": 1, "results": []}
```

What if we use the following work around for now?

```
diff --git a/ovn-fake-multinode-utils/playbooks/install-dependencies.yml b/ovn-fake-multinode-utils/playbooks/install-dependencies.yml                                                                                                       
index 5229252..608e21b 100644
--- a/ovn-fake-multinode-utils/playbooks/install-dependencies.yml
+++ b/ovn-fake-multinode-utils/playbooks/install-dependencies.yml
@@ -6,15 +6,24 @@
         name:
           - git
           - gcc
-          - openvswitch
           - python3-pyyaml
           - python3-virtualenv
           - python3-devel
           - containers-common
-        skip_broken: true
         state: present
       when: ansible_os_family == "RedHat"

+    # 'skip-broken' might not work with specific ansible versions
+    # so ignore_errors for packages that might not be there.
+    # https://github.com/ansible/ansible/commit/6bcb494f8306
+    - name: Install optional packages
+      ansible.builtin.package:
+        name:
+          - openvswitch
+        state: present
+      ignore_errors: true
+      when: ansible_os_family == "RedHat"
+
     - name: Install required packages
       ansible.builtin.package:
         name:
```

Comment (fnordahl) [reply ovn-fake-multinode-utils/playbooks/install-dependencies.yml]: Thx for testing @dceara, works for me, I'll include it in the PR

**Review by igsilya — APPROVED** [review]
Didn't test, but LGTM.

**Pull request merged**
